src/plugins/kibana/server/routes/api/ingest/register_post.js

```diff
@@ -1,26 +1,72 @@
 const Boom = require('boom');
 const _ = require('lodash');
-const {templateToPattern, patternToTemplate} = require('../../../lib/convert_pattern_and_template_name');
-const indexPatternSchema = require('../../../lib/schemas/resources/index_pattern_schema');
+const {ingestToPattern, patternToIngest} = require('../../../lib/convert_pattern_and_ingest_name');
+const ingestConfigSchema = require('../../../lib/schemas/resources/ingest_config_schema');
 const handleESError = require('../../../lib/handle_es_error');
 const { keysToCamelCaseShallow } = require('../../../lib/case_conversion');
 const createMappingsFromPatternFields = require('../../../lib/create_mappings_from_pattern_fields');
 const initDefaultFieldProps = require('../../../lib/init_default_field_props');
 
+function patternRollback(rootError, indexPatternId, boundCallWithRequest) {
+  const deleteParams = {
+    index: '.kibana',
+    type: 'index-pattern',
+    id: indexPatternId
+  };
+
+  return boundCallWithRequest('delete', deleteParams)
```

Comment:
Yeah, so the API creates an index pattern, an index template, and a pipeline. It should be an atomic operation, so if any one item fails, we need to rollback the previous steps. This `patternRollback` function gets called when, for instance, the template creation fails but the index pattern was already created successfully. If the index pattern rollback succeeds, `patternRollback` throws the original template creation error so that it gets returned to the client and they can fix their error. If the rollback fails, `patternRollback` throws an error which lets the user know that things have been left in a partially completed state, along with the original template creation error so they know why a rollback was attempted in the first place.

Let me know if that helps at all. I'd be happy to hop on zoom tomorrow and talk through it some more.